cmd/krel/cmd/sign_blobs.go

```diff
@@ -0,0 +1,218 @@
+		if err != nil {
+			return fmt.Errorf("creating a temporary directory to save the files to be signed: %w", err)
+		}
+
+		logrus.Infof("Getting a list of files to be signed from %s", args[0])
+		output, err := gcli.GSUtilOutput("ls", "-R", args[0])
+		if err != nil {
+			return fmt.Errorf("listing bucket contents: %w", err)
+		}
+
+		gcsClient := object.NewGCS()
+		for _, file := range strings.Fields(output) {
+			if strings.HasSuffix(file, ".sha256") || strings.HasSuffix(file, ".sha512") ||
+				strings.HasSuffix(file, ":") || strings.HasSuffix(file, ".docker_tag") ||
+				(strings.ContainsAny(file, "SHA256SUMS") && strings.ContainsAny(file, "README") &&
+					strings.ContainsAny(file, "Makefile")) {
+				continue
+			}
+			temp := strings.TrimPrefix(file, object.GcsPrefix)
+			err = gcsClient.CopyToLocal(file, tempDir)
+			if err != nil {
+				return fmt.Errorf("copying file to sign: %w", err)
+			}
+
+			bundle = append(bundle, signingBundle{
+				destinationPathToCopy: filepath.Dir(temp),
+				fileToSign:            filepath.Base(temp),
+				fileLocalLocation:     fmt.Sprintf("%s/%s", tempDir, filepath.Base(temp)),
+			})
+		}
+	} else {
+		// Local files
+		for _, arg := range args {
+			bundle = append(bundle, signingBundle{
+				fileLocalLocation: arg,
+				fileToSign:        filepath.Base(arg),
+			})
+		}
+	}
+
+	t := throttler.New(int(signOpts.maxWorkers), len(args))
+	for _, fileBundle := range bundle {
+		go func(fileBundle signingBundle) {
+			logrus.Infof("Signing %s...", fileBundle.fileToSign)
+			signerOpts := sign.Default()
+			signerOpts.Verbose = signOpts.verbose
+			signerOpts.Timeout = signOpts.timeout
+			signerOpts.PrivateKeyPath = signBlobOpts.privateKeyPath
+			signerOpts.PublicKeyPath = signBlobOpts.publicKeyPath
+
+			signerOpts.OutputCertificatePath = fmt.Sprintf("%s/%s.cert", signBlobOpts.outputPath, fileBundle.fileToSign)
+			signerOpts.OutputSignaturePath = fmt.Sprintf("%s/%s.sig", signBlobOpts.outputPath, fileBundle.fileToSign)
+			if signBlobOpts.outputPath == "" {
+				signerOpts.OutputCertificatePath = fmt.Sprintf("%s.cert", fileBundle.fileLocalLocation)
+				signerOpts.OutputSignaturePath = fmt.Sprintf("%s.sig", fileBundle.fileLocalLocation)
+			}
+
+			signer := sign.New(signerOpts)
+			_, err := signer.SignFile(fileBundle.fileLocalLocation)
+			if err != nil {
+				t.Done(fmt.Errorf("signing the file %s: %w", fileBundle.fileLocalLocation, err))
+				return
+			}
+			t.Done(nil)
+		}(fileBundle)
+
+		if t.Throttle() > 0 {
+			break
+		}
+	}
+	if err := t.Err(); err != nil {
+		return fmt.Errorf("signing the blobs: %w", err)
+	}
+
+	if isGCSBucket {
+		logrus.Info("Copying Certificates and Signatures back to the bucket...")
+		for _, fileBundle := range bundle {
+			certSignFiles := fmt.Sprintf("%s/%s*", signBlobOpts.outputPath, fileBundle.fileToSign)
+			if signBlobOpts.outputPath == "" {
+				certSignFiles = fmt.Sprintf("%s*", fileBundle.fileLocalLocation)
+			}
```

Comment:
I have some concerns that this might override binaries in some cases. For example, what if you have two binaries named `kubectl` and `kubectl-convert`? In that case, you're going to copy:
- `kubectl*`
- `kubectl-convert*`

The first one might copy:
- `kubectl.sig`
- `kubectl.cert`
- `kubectl-convert` <-- this is another binary itself
- `kubectl-convert.sig` <-- this is a signature of another binary
- `kubectl-convert.cert` <-- this is a certificate of another binary

This should be a no-op, but I'm wondering if we should do this at all or if this can cause any failure.

Reply:
great catch, lets filter with .sig and .cert?

Reply:
> lets filter with .sig and .cert?

Yes, that makes sense to me.

Reply:
done